### PR TITLE
Modbus driver

### PR DESCRIPTION
**Pull request by pkowalsk1**

### CODE REVIEW

File: onrobot_rg2_description/urdf/onrobot.urdf.xacro [hunk 1]
```diff
@@ -10,6 +10,6 @@
   <xacro:gripper.onrobot_rg2 
     parent_link="quick_changer_link" 
     xyz="0.0 0.0 0.015" 
-    rpy="${-pi/2} ${-pi/2} 0.0" />
+    rpy="${-pi/2} ${-pi/2} ${-pi/2}" />
```
Comment (KmakD) [inline onrobot_rg2_description/urdf/onrobot.urdf.xacro]: Can't this be simplified to just?
```suggestion
    rpy="0.0 ${-pi/2} 0.0" />
```

File: onrobot_rg2_driver/launch/bringup.launch [hunk 2]
```diff
@@ -1,3 +1,15 @@
 <launch>
-  <node name="rg2_driver_node" pkg="onrobot_rg2_driver" type="rg2_driver.py" output="screen" />
+
+  <arg name="ur_robot_ip" default="10.15.20.4" />
+  <arg name="base_link_frame" default="rg2_base_link" />
+  <arg name="gripper_joint_name" default="rg2_gripper_joint" />
+  <arg name="service_call_timeout_sec" default="10" />
```
Comment (KmakD) [inline onrobot_rg2_driver/launch/bringup.launch]:
```suggestion
  <arg name="service_call_timeout_sec" default="10.0" />
```

File: onrobot_rg2_driver/scripts/husarion_ext_control.script [hunk 3]
```diff
@@ -0,0 +1,1539 @@
...
+  $ 4 "Thread_1"
+  thread Thread_1():
+    $ 5 "Script: husarion_rg2_driver.script"
+    modbus_add_signal("10.42.0.49", 255, 16, 3, "joint_out")
+    modbus_add_signal("10.42.0.49", 255, 17, 1, "rg_busy_out")
+    modbus_add_signal("10.42.0.49", 255, 18, 1, "grip_detected_out")
+    
+    modbus_add_signal("10.42.0.49", 255, 20, 0, "action_trigger_in")
+    modbus_add_signal("10.42.0.49", 255, 128, 2, "joint_in")
+    modbus_add_signal("10.42.0.49", 255, 129, 2, "force_in")
```
Comment (KmakD) [inline onrobot_rg2_driver/scripts/husarion_ext_control.script]: Will this work with UR with different ip's?

File: onrobot_rg2_driver/scripts/husarion_ext_control.txt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+ Program
+   Variables Setup
+   Robot Program
+     Control by 10.42.0.1
```
Comment (KmakD) [inline onrobot_rg2_driver/scripts/husarion_ext_control.txt]: Same as above

File: onrobot_rg2_driver/scripts/husarion_rg2_driver.script [hunk 5]
```diff
@@ -0,0 +1,29 @@
+modbus_add_signal("10.42.0.49", 255, 16, 3, "joint_out")
+modbus_add_signal("10.42.0.49", 255, 17, 1, "rg_busy_out")
+modbus_add_signal("10.42.0.49", 255, 18, 1, "grip_detected_out")
+
+modbus_add_signal("10.42.0.49", 255, 20, 0, "action_trigger_in")
+modbus_add_signal("10.42.0.49", 255, 128, 2, "joint_in")
+modbus_add_signal("10.42.0.49", 255, 129, 2, "force_in")
```
Comment (KmakD) [inline onrobot_rg2_driver/scripts/husarion_rg2_driver.script]: same as above

File: onrobot_rg2_driver/scripts/husarion_rg2_watchdog.script [hunk 6]
```diff
@@ -0,0 +1,10 @@
+modbus_add_signal("10.42.0.49", 255, 19, 1, "watchdog_out")
```
Comment (KmakD) [inline onrobot_rg2_driver/scripts/husarion_rg2_watchdog.script]: same as above

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    MIN_RG2_FORCE: float = 3.0  # [N]
+    MAX_RG2_FORCE: float = 40.0  # [N]
+    MIN_RG2_RANGE: float = 0.0  # [N]
+    MAX_RG2_RANGE: float = 100.0  # [N]
+    TIMEOUT: float = 10.0  # [s]
+
+    def __init__(self) -> None:
+        rospy.init_node("rg2_driver_node", anonymous=False)
+
+        ur_robot_ip = rospy.get_param('~ur_robot_ip', '10.42.0.49')
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
        ur_robot_ip = rospy.get_param('~ur_robot_ip', '10.15.20.4')
```
according to readme this is deafult

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        self._joint_state_msg = JointState()
+        self._joint_state_msg.header.frame_id = frame_id
+        self._joint_state_msg.name = [joint_name]
+
+        rate = rospy.Rate(0.5)
+        while not self._rg2_modbus.watchdog.is_active():
+            rospy.loginfo_throttle(5.0, 'Waiting for URscript Modbus watchdog activation...')
+            rate.sleep()
+
+        self._joint_publisher = rospy.Publisher("/joint_states", JointState, queue_size=1)
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py:43]: Won't this affect `/joint_states` published by Panther when working with robot?

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    TIMEOUT: float = 10.0  # [s]
+
+    def __init__(self) -> None:
+        rospy.init_node("rg2_driver_node", anonymous=False)
+
+        ur_robot_ip = rospy.get_param('~ur_robot_ip', '10.42.0.49')
+        frame_id = rospy.get_param('~base_link_frame', 'rg2_base_link')
+        joint_name = rospy.get_param('~gripper_joint_name', 'rg2_gripper_joint')
+        self.service_call_timeout_sec = rospy.get_param('~service_call_timeout_sec', 10.0)
+        self._rg2_force_range = (3.0, 40.0)  # [N]
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
```
not used

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            "rg2/set_gripper_width", GripperState, self._set_grip_state
+        )
+
+        rospy.Timer(rospy.Duration(0.05), self._pub_joint_state_cb)
+
+        rospy.loginfo(f'[{rospy.get_name()}] Node started')
+
+    def _set_grip_state(self, req: GripperStateRequest) -> GripperStateResponse:
+        try:
+            with self._lock:
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
        with self._lock:
            try:        
```
I'd rather do it this way

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                    raise ValueError(
+                        f'Force value must be between {self.MIN_RG2_FORCE} and {self.MAX_RG2_FORCE} [N]'
+                    )
+                if not self.MIN_RG2_RANGE <= req.width <= self.MAX_RG2_RANGE:
+                    raise ValueError(
+                        f'Width value must be between {self.MIN_RG2_RANGE} and {self.MAX_RG2_RANGE} [mm]'
+                    )
+
+                rospy.loginfo(
+                    f'[{rospy.get_name()}] Sending gripp request:\n'
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
                    f'[{rospy.get_name()}] Sending gripper request:\n'
```

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                    f'\t - width: {req.width}\n\t - force: {req.force}'
+                )
+
+                self._rg2_modbus.send_gripp_req(req.width, req.force)
+                grip_detected = self._wait_for_grip_completion(self.TIMEOUT)
+
+                rospy.loginfo(f'[{rospy.get_name()}] Grip action succes!')
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
                rospy.loginfo(f'[{rospy.get_name()}] Gripper action success!')
```

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+                return GripperStateResponse(True, grip_detected, None)
+
+        except (ValueError, TimeoutError, ModbusError, ConnectionError) as e:
+            rospy.logwarn(f'[{rospy.get_name()}] Grip action failed: {str(e)}')
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
            rospy.logwarn(f'[{rospy.get_name()}] Gripper action failure: {str(e)}')
```
or in above action succeeded, to keep it consistent

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            return GripperStateResponse(False, False, str(e))
+
+    def _wait_for_grip_completion(self, timeout: float) -> bool:
+        start_time = rospy.Time.now()
+        rate = rospy.Rate(20)
+
+        while not self._rg2_modbus.querry_rg_busy_flag():
+            if (rospy.Time.now() - start_time).secs > timeout / 10:
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
        start_time = rospy.get_time()
        rate = rospy.Rate(20)

        while not self._rg2_modbus.querry_rg_busy_flag():
            if (rospy.get_time() - start_time) > timeout / 10:
```

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+                # The preset position of the gripper is the same as 
+                # the actual position before the service was sent
+                return self._rg2_modbus.querry_grip_detected_flag()
+            rate.sleep()
+
+        rospy.loginfo(f'[{rospy.get_name()}] Request send succesful.')
+
+        while self._rg2_modbus.querry_rg_busy_flag():
+            if (rospy.Time.now() - start_time).secs > timeout:
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
            if (rospy.get_time() - start_time) > timeout:
```

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+        rate = rospy.Rate(20)
+
+        while not self._rg2_modbus.querry_rg_busy_flag():
+            if (rospy.Time.now() - start_time).secs > timeout / 10:
+                # The preset position of the gripper is the same as 
+                # the actual position before the service was sent
+                return self._rg2_modbus.querry_grip_detected_flag()
+            rate.sleep()
+
+        rospy.loginfo(f'[{rospy.get_name()}] Request send succesful.')
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
        rospy.loginfo(f'[{rospy.get_name()}] Request sent successfully.')
```

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+        while self._rg2_modbus.querry_rg_busy_flag():
+            if (rospy.Time.now() - start_time).secs > timeout:
+                raise TimeoutError('Grip action failed: timeout')
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
                raise TimeoutError('Gripper action failed: timeout')
```

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+            rate.sleep()
+
+        return self._rg2_modbus.querry_grip_detected_flag()
+
+    def _pub_joint_state_cb(self, *args) -> None:
+        try:
+            self._joint_state_msg.header.stamp = rospy.Time.now()
+
+            rg_width = self._rg2_modbus.querry_gripper_width()
+            joint_state = self._map(rg_width, 0.0, 101.0, -0.44, 1.0)
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py:106]:
```suggestion
            joint_state = self._map(rg_width, 0.0, 100.0, -0.44, 1.0)
```
Shouldn't be in range <0,100>?

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+            self._joint_state_msg.position = [joint_state]
+
+            self._joint_publisher.publish(self._joint_state_msg)
+
+        except ConnectionError as e:
+            rospy.logerr_throttle(20.0, str(e))
+
+    @staticmethod
+    def _map(x: float, in_from: float, in_to: float, out_from: float, out_to: float) -> float:
+        return (x - in_from) / (in_to - in_from) * (out_to - out_from) + out_from
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]: I would do: in_min, in_max, out_min, out_max

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+
+
+if __name__ == "__main__":
+    try:
+        rg2_driver_node = Rg2DriverNode()
+        rospy.spin()
+    except rospy.ROSInterruptException:
+        pass
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
def main():
    rg2_driver_node = Rg2DriverNode('rg2_driver_node')
    rospy.spin()


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass
```
and also add name to class definition

File: onrobot_rg2_driver/src/rg2_modbus.py [hunk 21]
```diff
@@ -0,0 +1,96 @@
+from dataclasses import dataclass
+import modbus_tk.defines as cst
+from modbus_tk import modbus_tcp
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_modbus.py]: wrong order I think

File: onrobot_rg2_driver/src/rg2_modbus.py [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+class Watchdog:
+    WATCHDOG_TIMEOUT = rospy.Duration(0.2)
+
+    def __init__(self, master: modbus_tcp.TcpMaster) -> None:
+        self._master = master
+        self._prev_state = False
+        self._last_state_change_time = None
+
+        self.watchdog_thread = Thread(target=self._update_watchdog_state, name='watchdog_thread')
+        self.watchdog_thread.start()
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_modbus.py]:
```suggestion
        self._watchdog_thread = Thread(target=self._update_watchdog_state, name='watchdog_thread')
        self._watchdog_thread.start()
```

File: onrobot_rg2_driver/src/rg2_modbus.py [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+        rospy.sleep(self.WATCHDOG_TIMEOUT)
+
+    def _update_watchdog_state(self):
+        rate = rospy.Rate(20)
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_modbus.py]:
```suggestion
        rate = rospy.Rate(20.0)
```

File: onrobot_rg2_description/launch/view_rg2.launch [hunk 24]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
+<launch>
+    <arg name="publish_robot_state" default="true" />
+
+    <param command="$(find xacro)/xacro $(find onrobot_rg2_description)/urdf/onrobot.urdf.xacro"
+        name="robot_description" />
+
+    <node name="joint_state_publisher_gui" pkg="joint_state_publisher_gui" type="joint_state_publisher_gui" if="$(arg publish_robot_state)" />
+    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
+    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find onrobot_rg2_description)/config/view_rg2.rviz" required="true" />
```
Comment (KmakD) [inline onrobot_rg2_description/launch/view_rg2.launch]: usual order for nodes (as for panther_ros pkgs) is pkg, type, name

**Review by KmakD — CHANGES_REQUESTED** [review]

File: onrobot_rg2_driver/src/rg2_driver_node.py (hunk 8 above)
Comment (pkowalsk1) [reply onrobot_rg2_driver/src/rg2_driver_node.py:43]: This is normal practice. Joint states are parsed based on strings.

File: onrobot_rg2_driver/src/rg2_driver_node.py (hunk 18 above)
Comment (pkowalsk1) [reply onrobot_rg2_driver/src/rg2_driver_node.py:106]: The gripper has some offset and this results in better accuracy :D

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+
+                rospy.loginfo(
+                    f'[{rospy.get_name()}] Sending gripper request:\n'
+                    f'\t - width: {req.width}\n\t - force: {req.force}'
+                )
+
+                self._rg2_modbus.send_gripp_req(req.width, req.force)
+                grip_detected = self._wait_for_grip_completion(self._service_call_timeout)
+
+                rospy.loginfo(f'[{rospy.get_name()}] Gripper action succes!')
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
                rospy.loginfo(f'[{rospy.get_name()}] Gripper action success!')
```
or succeeded

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 26]
```diff
@@ -0,0 +1,128 @@
...
+            except (ValueError, TimeoutError, ModbusError, ConnectionError) as e:
+                rospy.logwarn(f'[{rospy.get_name()}] Gripper action failed: {str(e)}')
+                return GripperStateResponse(False, False, str(e))
+
+    def _wait_for_grip_completion(self, timeout: Duration) -> bool:
+        start_time = rospy.Time.now()
+        rate = rospy.Rate(20)
+
+        while not self._rg2_modbus.querry_rg_busy_flag():
+            if rospy.Time.now() - start_time > timeout / 10:
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
            if rospy.Time.now() - start_time > timeout / 10.0:
```

**Review by KmakD — CHANGES_REQUESTED** [review]

File: onrobot_rg2_driver/src/rg2_driver_node.py [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+
+                return GripperStateResponse(True, grip_detected, None)
+
+            except (ValueError, TimeoutError, ModbusError, ConnectionError) as e:
+                rospy.logwarn(f'[{rospy.get_name()}] Gripper action failed: {str(e)}')
+                return GripperStateResponse(False, False, str(e))
+
+    def _wait_for_grip_completion(self, timeout: Duration) -> bool:
+        start_time = rospy.Time.now()
+        rate = rospy.Rate(20)
```
Comment (KmakD) [inline onrobot_rg2_driver/src/rg2_driver_node.py]:
```suggestion
        rate = rospy.Rate(20.0)
```

**Review by KmakD — CHANGES_REQUESTED** [review]

**Review by KmakD — APPROVED** [review]

**Pull request merged**